module/zfs/zio.c

```diff
@@ -3252,9 +3252,31 @@ zio_ddt_write(zio_t *zio)
 	ASSERT(!(zio->io_bp_override && (zio->io_flags & ZIO_FLAG_RAW)));
 
 	ddt_enter(ddt);
-	dde = ddt_lookup(ddt, bp, B_TRUE);
+	if (spa->spa_dedup_table_quota &&
+	    ddt_get_ddt_dsize(spa) >= spa->spa_dedup_table_quota) {
+		/* Over the desired DDT size, do not grow the table */
+		dde = ddt_lookup(ddt, bp, B_FALSE);
+	} else {
+		/* It's ok to grow the table. */
+		dde = ddt_lookup(ddt, bp, B_TRUE);
+	}
+	if (dde == NULL) {
+		zp->zp_dedup = B_FALSE;
+		BP_SET_DEDUP(bp, B_FALSE);
+		ASSERT(!BP_GET_DEDUP(bp));
```

Comment:
Is this assertion because are are worried that BP_SET_DEDUP() doesn't do what it says?  Seems like overkill to me.

Reply:
I believe I copied the assertion from existing code.  Probably a paste error on my part to have it, though.